core/src/main/java/io/questdb/griffin/engine/groupby/StableAwareUtf8StringHolder.java

```diff
@@ -0,0 +1,172 @@
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Specialized flyweight used in some {@link io.questdb.griffin.engine.functions.GroupByFunction}s. Unlike sinks
+ * this holder cannot append to the underlying buffer. When you set a value then any previous value is cleared.
+ * <p>
+ * It can hold either a pointer to a UTF-8 sequence or a UTF-8 sequence itself. When a sequence to be held is
+ * direct, then the Holder stores only a pointer to the char sequence and its length. Otherwise, it stores the char
+ * sequence itself.
+ * <br>
+ * The information about whether a stored sequence is direct or not is not stored in the header of the Holder. Instead, the
+ * top bit of the pointer returned by {@link #ptr()} is used to store this information. This is done to save space in the
+ * header and to avoid the need to store this information separately. Thus, the value returned by {@link #ptr()} cannot
+ * be used as a pointer directly and should be treated as an opaque value.
+ * <p>
+ * Uses provided {@link GroupByAllocatorImpl} to allocate the underlying buffer. Grows the buffer when needed.
+ * <p>
+ * Buffer layout is the following:
+ * <pre>
+ * | capacity (in bytes) |  size (in bytes)  | is ascii | byte array or pointer |
+ * +---------------------+-------------------+----------+-----------------------+
+ * |       4 bytes       |      4 bytes      | 1 byte   |           -           |
+ * +---------------------+-------------------+----------+-----------------------+
+ * </pre>
+ */
+public class StableAwareUtf8StringHolder implements Utf8Sequence {
+    private static final long HEADER_SIZE = 2 * Integer.BYTES + Byte.BYTES;
+    private static final long IS_ASCII_OFFSET = 2 * Integer.BYTES;
+    private static final int MIN_CAPACITY = 8; // 8 bytes = enough to store a pointer
+    private static final long SIZE_OFFSET = Integer.BYTES;
+    private GroupByAllocator allocator;
+    private AsciiCharSequence asciiCharSequence;
+    private boolean direct;
+    private long ptr;
+
+    @Override
+    public @NotNull CharSequence asAsciiCharSequence() {
+        if (asciiCharSequence == null) {
+            asciiCharSequence = new AsciiCharSequence();
+        }
+        asciiCharSequence.of(this);
+        return asciiCharSequence;
+    }
+
+    @Override
+    public byte byteAt(int index) {
+        if (direct) {
+            // we could cache the direct pointer, but then we would need to invalidate it when the pointer changes
+            // and we assume of() is called more frequently than charAt()
+            long directPtr = Unsafe.getUnsafe().getLong(ptr + HEADER_SIZE);
+            assert directPtr != 0;
+            return Unsafe.getUnsafe().getByte(directPtr + index);
+        } else {
+            return Unsafe.getUnsafe().getByte(ptr + HEADER_SIZE + index);
+        }
+    }
+
+    public void clearAndSet(@Nullable Utf8Sequence us) {
+        clear();
+        if (us instanceof StableDirectSequence) {
```

Comment:
We could get rid of this `instanceof`, but that would require adding a few methods to the basic `Utf8Sequence` interface: ```java
interface Utf8Sequence {

    boolean isDirect();

    // can be called only is isDirect is true
    long ptr();

    // returns true for on-heap immutable strings (Utf8String)
    // and off-heap strings backed with mmapped memory
    boolean isStable();
    // ...
}
```

Then this check would look like this:
```java
    public void clearAndSet(@Nullable Utf8Sequence us) {
        clear();
        if (us.isDirect() && us.isStable()) {
            direct = true;
            checkCapacity(4); // pointer is 8 bytes = 4 chars
            Unsafe.getUnsafe().putLong(ptr + HEADER_SIZE, us.ptr());
            Unsafe.getUnsafe().putInt(ptr + SIZE_OFFSET, us.size());
            Unsafe.getUnsafe().putBoolean(null, ptr + IS_ASCII_OFFSET, us.isAscii());
```

@bluestreak01 do you think it's worth it to get rid of the problematic `instanceof` just for UTF-8 sequences while char sequence optimization still uses `instanceof`?

Reply:
yep, i think it is worth removing `instaceOf`

Reply:
Got rid of `instanceof` interface checks here: 5dd6a05f24aad8b0b85d968b5eeb6aa78e69b87c

For `Utf8Sequence`, I introduced `ptr()` and `isStable()` methods on the base interface, so that we check `us.isStable()` instead of dealing with `instanceof`.

For `CharSequence`, I deleted the marker interface (`StableCharSequence`) and only left the marker class (`StableDirectString`)